nifi-commons/nifi-sensitive-property-provider/src/main/java/org/apache/nifi/properties/ApplicationPropertiesProtector.java

```diff
@@ -0,0 +1,348 @@
+        return allSensitiveKeys.stream().filter(k -> StringUtils.isNotBlank(getProperty(k))).collect(Collectors.toList());
+    }
+
+    @Override
+    public boolean hasProtectedKeys() {
+        final List<String> sensitiveKeys = getSensitivePropertyKeys();
+        for (String k : sensitiveKeys) {
+            if (isPropertyProtected(k)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    @Override
+    public Map<String, String> getProtectedPropertyKeys() {
+        final List<String> sensitiveKeys = getSensitivePropertyKeys();
+
+        // This is the Java 8 way, but can likely be optimized (and not sure of correctness)
+        // Map<String, String> protectedProperties = sensitiveKeys.stream().filter(key ->
+        // getProperty(getProtectionKey(key)) != null).collect(Collectors.toMap(Function.identity(), key ->
+        // getProperty(getProtectionKey(key))));
+
+        // Groovy
+        // Map<String, String> groovyProtectedProperties = sensitiveKeys.collectEntries { key ->
+        // [(key): getProperty(getProtectionKey(key))] }.findAll { k, v -> v }
+
+        // Traditional way
+        final Map<String, String> traditionalProtectedProperties = new HashMap<>();
+        for (final String key : sensitiveKeys) {
+            final String protection = getProperty(getProtectionKey(key));
+            if (StringUtils.isNotBlank(protection) && StringUtils.isNotBlank(getProperty(key))) {
+                traditionalProtectedProperties.put(key, protection);
+            }
+        }
+
+        return traditionalProtectedProperties;
+    }
+
+    @Override
+    public Set<String> getProtectionSchemes() {
+        return new HashSet<>(getProtectedPropertyKeys().values());
+    }
+
+    @Override
+    public boolean isPropertySensitive(final String key) {
+        // If the explicit check for ADDITIONAL_SENSITIVE_PROPERTIES_KEY is not here, this will loop infinitely
+        return key != null && !key.equals(getAdditionalSensitivePropertiesKeysName()) && getSensitivePropertyKeys().contains(key.trim());
+    }
+
+    /**
+     * Returns true if the property identified by this key is considered protected in this instance of {@code NiFiProperties}.
+     * The property value is protected if the key is sensitive and the sibling key of key.protected is present.
+     *
+     * @param key the key
+     * @return true if it is currently marked as protected
+     * @see ApplicationPropertiesProtector#getSensitivePropertyKeys()
+     */
+    @Override
+    public boolean isPropertyProtected(final String key) {
+        return key != null && isPropertySensitive(key) && !StringUtils.isBlank(getProperty(getProtectionKey(key)));
+    }
+
+    @Override
+    public U getUnprotectedProperties() throws SensitivePropertyProtectionException {
+        if (hasProtectedKeys()) {
+            logger.info("There are {} protected properties of {} sensitive properties",
```

Comment:
Recommend streamlining the log message and changing it to debug
```suggestion
            logger.debug("Protected Properties [{}] Sensitive Properties [{}]",
```